tests/integration/rollup/rollup_test.go

```diff
@@ -0,0 +1,139 @@
+	})
+	require.NoError(t, err)
+
+	// query the user's ETH balance and assert it's equal to the deposit amount
+	require.Equal(t, depositAmount, bankKeeper.GetBalance(integrationApp.Context(), userAddr, rolluptypes.ETH).Amount.BigInt())
+
+	// try to withdraw more than deposited and assert error
+	_, err = integrationApp.RunMsg(&rolluptypes.MsgInitiateWithdrawal{
+		Sender:   userAddr.String(),
+		Target:   l1WithdrawalAddr,
+		Value:    math.NewIntFromBigInt(depositAmount).Add(math.OneInt()),
+		GasLimit: new(big.Int).SetUint64(100_000_000).Bytes(),
+		Data:     []byte{0x01, 0x02, 0x03},
+	})
+	require.Error(t, err)
+
+	// send a successful MsgInitiateWithdrawal
+	_, err = integrationApp.RunMsg(&rolluptypes.MsgInitiateWithdrawal{
+		Sender:   userAddr.String(),
+		Target:   l1WithdrawalAddr,
+		Value:    math.NewIntFromBigInt(depositAmount),
+		GasLimit: new(big.Int).SetUint64(100_000_000).Bytes(),
+		Data:     []byte{0x01, 0x02, 0x03},
+	})
+	require.NoError(t, err)
+
+	// query the user's ETH balance and assert it's zero
+	require.Equal(t, math.ZeroInt(), bankKeeper.GetBalance(integrationApp.Context(), userAddr, rolluptypes.ETH).Amount)
+}
+
+func setupIntegrationApp(t *testing.T) (*integration.App, bankkeeper.BaseKeeper) {
+	encodingCfg := moduletestutil.MakeTestEncodingConfig(auth.AppModuleBasic{}, bank.AppModuleBasic{}, rollup.AppModuleBasic{})
+	keys := storetypes.NewKVStoreKeys(authtypes.StoreKey, banktypes.StoreKey, rolluptypes.StoreKey)
+	authority := authtypes.NewModuleAddress("gov").String()
+
+	logger := log.NewTestLogger(t)
+
+	cms := integration.CreateMultiStore(keys, logger)
+
+	accountKeeper := authkeeper.NewAccountKeeper(
+		encodingCfg.Codec,
+		runtime.NewKVStoreService(keys[authtypes.StoreKey]),
+		authtypes.ProtoBaseAccount,
+		map[string][]string{rolluptypes.ModuleName: {authtypes.Minter, authtypes.Burner}},
+		addresscodec.NewBech32Codec("cosmos"),
+		"cosmos",
+		authority,
+	)
+	bankKeeper := bankkeeper.NewBaseKeeper(
+		encodingCfg.Codec,
+		runtime.NewKVStoreService(keys[banktypes.StoreKey]),
+		accountKeeper,
+		map[string]bool{},
+		authority,
+		logger,
+	)
+	rollupKeeper := rollupkeeper.NewKeeper(
+		encodingCfg.Codec,
+		runtime.NewKVStoreService(keys[rolluptypes.StoreKey]),
+		bankKeeper,
+	)
+
+	authModule := auth.NewAppModule(encodingCfg.Codec, accountKeeper, authsims.RandomGenesisAccounts, nil)
+	bankModule := bank.NewAppModule(encodingCfg.Codec, bankKeeper, accountKeeper, nil)
+	rollupModule := rollup.NewAppModule(encodingCfg.Codec, rollupKeeper)
+
+	integrationApp := integration.NewIntegrationApp(
+		sdk.NewContext(cms, cmtproto.Header{}, false, logger),
+		logger,
+		keys,
+		encodingCfg.Codec,
+		map[string]appmodule.AppModule{
+			authtypes.ModuleName:   authModule,
+			banktypes.ModuleName:   bankModule,
+			rolluptypes.ModuleName: rollupModule,
+		},
+	)
+	rolluptypes.RegisterMsgServer(integrationApp.MsgServiceRouter(), rollupkeeper.NewMsgServerImpl(rollupKeeper))
+
+	return integrationApp, bankKeeper
```

Comment:
Rather than using the raw keeper, might be better to only expose the query client. https://github.com/cosmos/cosmos-sdk/blob/fb7775ad0c1cd70f6b9a3305a3ebde9766ad0454/tests/integration/bank/keeper/deterministic_test.go#L139-L140

Reply:
Good point, I'll look into changing this to use the query client